iree/compiler/Dialect/Stream/IR/StreamOpFolders.cpp

```diff
@@ -653,7 +661,15 @@ void TensorImportOp::getCanonicalizationPatterns(
 //===----------------------------------------------------------------------===//
 
 OpFoldResult TensorExportOp::fold(ArrayRef<Attribute> operands) {
-  // TODO(benvanik): if operand comes from import then fold.
+  // If operand comes from an import with the same affinity and size then fold.
+  // Different affinities may indicate importing from one device or queue and
+  // exporting to a different device or queue.
+  // We assume that differing encodings and shapes are compatible.
+  auto importOp = source().getDefiningOp<TensorImportOp>();
+  if (importOp && affinity() == importOp.affinity() &&
+      source_size() == importOp.result_size()) {
+    return importOp.source();
+  }
```

Comment:
one test failure had different behavior after folding all of this to `return %arg0 : !hal.buffer_view`:

```mlir
func @Flatten__3x3x3x3x3__f32__uniform(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub, iree.reflection = {iree.abi = "{\22a\22:[[\22ndarray\22,\22f32\22,5,3,3,3,3,3]],\22r\22:[[\22ndarray\22,\22f32\22,2,3,81]],\22v\22:1}"}} {
  %c972 = arith.constant 972 : index
  %0 = stream.tensor.import %arg0 : !hal.buffer_view -> tensor<3x3x3x3x3xf32> in !stream.resource<external>{%c972}
  %1 = stream.tensor.export %0 : tensor<3x81xf32> in !stream.resource<external>{%c972} -> !hal.buffer_view
  return %1 : !hal.buffer_view
}
```

without the folding, that becomes:
```mlir
// -----// IR Dump After mlir::iree_compiler::IREE::HAL::`anonymous-namespace'::ConvertToHALPass //----- //
#device_target_cpu = #hal.device.target<"cpu", {executable_targets = [#hal.executable.target<"llvm", "embedded-elf-x86_64", {data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>]}>
module attributes {hal.device.targets = [#device_target_cpu]}  {
  func @Flatten__3x3x3x3x3__f32__uniform(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub, iree.reflection = {iree.abi = "{\22a\22:[[\22ndarray\22,\22f32\22,5,3,3,3,3,3]],\22r\22:[[\22ndarray\22,\22f32\22,2,3,81]],\22v\22:1}"}} {
    %c972 = arith.constant 972 : index
    %buffer = hal.buffer_view.buffer<%arg0 : !hal.buffer_view> : !hal.buffer
    %c50331680_i32 = arith.constant 50331680 : i32
    %c1_i32 = arith.constant 1 : i32
    %c3 = arith.constant 3 : index
    %c3_0 = arith.constant 3 : index
    %c3_1 = arith.constant 3 : index
    %c3_2 = arith.constant 3 : index
    %c3_3 = arith.constant 3 : index
    hal.buffer_view.assert<%arg0 : !hal.buffer_view> message("tensor") shape([%c3, %c3_0, %c3_1, %c3_2, %c3_3]) type(%c50331680_i32) encoding(%c1_i32)
    %device = hal.ex.shared_device : !hal.device
    %allocator = hal.device.allocator<%device : !hal.device> : !hal.allocator
    hal.buffer.assert<%buffer : !hal.buffer> message("tensor") allocator(%allocator : !hal.allocator) minimum_length(%c972) type(DeviceVisible) usage("Transfer|Dispatch")
    %c3_4 = arith.constant 3 : index
    %c81 = arith.constant 81 : index
    %c1_i32_5 = arith.constant 1 : i32
    %c50331680_i32_6 = arith.constant 50331680 : i32
    %view = hal.buffer_view.create buffer(%buffer : !hal.buffer) shape([%c3_4, %c81]) type(%c50331680_i32_6) encoding(%c1_i32_5) : !hal.buffer_view
    return %view : !hal.buffer_view
  }
}
```

this was the input IR:
```mlir
#loc0 = loc(unknown)
module  {
  func @Flatten__3x3x3x3x3__f32__uniform(%arg0: !hal.buffer_view loc(unknown)) -> !hal.buffer_view attributes {iree.abi = "{\22a\22:[[\22ndarray\22,\22f32\22,5,3,3,3,3,3]],\22r\22:[[\22ndarray\22,\22f32\22,2,3,81]],\22v\22:1}"} {
    %0 = hal.tensor.cast %arg0 : !hal.buffer_view -> tensor<3x3x3x3x3xf32> loc(#loc0)
    %1 = linalg.tensor_collapse_shape %0 [[0], [1, 2, 3, 4]] : tensor<3x3x3x3x3xf32> into tensor<3x81xf32> loc(#loc1)
    %2 = hal.tensor.cast %1 : tensor<3x81xf32> -> !hal.buffer_view loc(#loc0)
    return %2 : !hal.buffer_view loc(#loc0)
  } loc(#loc0)
} loc(#loc0)
#loc1 = loc(fused["Reshape:", callsite("model_4/flatten_4/Reshape@__inference_<lambda>_73"("/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py":1083:0) at callsite("/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py":60:0 at callsite("./iree/integrations/tensorflow/e2e/keras/layers/layers_test.py":512:0 at callsite("/usr/local/lib/python3.7/dist-packages/tensorflow/python/autograph/core/function_wrappers.py":117:0 at callsite("./iree/integrations/tensorflow/e2e/keras/layers/layers_test.py":512:0 at callsite("/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py":1125:0 at callsite("/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/def_function.py":672:0 at callsite("/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py":1143:0 at callsite("/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py":3402:0 at "/usr/local/lib/python3.7/dist-packages/tensorflow/python/eager/function.py":3557:0)))))))))])
```

Reply:
Oh, I think I misinterpreted "so it's always safe to fold export->import even if the encoding and dims don't match" to also mean the same for import->export. Maybe this is better now?